### ISSUE TITLE
Incorrect DB hook dirty field in case of concurrent modification

### ISSUE THREAD

**Issue by ylopotun** [orientechnologies/orientdb#4192]
We encountered a DB hook-related issue. When we try to set a vertex property via remote connection and the vertex object is not of the latest version, the DB hook’s dirty fields list mistakenly contains the property that was changed in between the current object’s version and the latest, although it wasn’t meant to be changed in the current operation. 
See the attached unit test for better understanding.
For better understanding, consider the following example:
1) Thread 1 (plocal connection) created Vertex 1 and sets its “name” property value to “name1”
2) Thread 2 (remote connection) loads Vertex 1
3) Thread 1 updates Vertex 1 “name” property value to “name2”
4) Thread 2 tries to set Vertex 1 “description” property value to “description1”. The operation fails with OConcurrentModificationExcpetion, but the DB Hook gets fired indicating the change of 2 property values:
“name” – from “name2” to “name1” (which is incorrect – this property value was not meant to be changed)
“description” – from null to “description1”
5) Thread 2 reloads Vertex 1 and retries setting its “description” property value to “description1”. This time operation succeeds. The DB Hook gets fired indicating the change of 1 property value:
“description” – from null to “description1”

This results in the modules that rely on DB notifications to have the wrong value of the “name” property.
So, either in step 4 the DB Hook should not notify about “name” property value change to the previous one or in step 5 it should again notify about “name” property value change, this time to the current one.

**Comment by tglman** [comment]
hi @ylopotun 

I got your code and run the tests, i see your problem, so i changed a bit the implementation of the hook in this way :
```
public class UpdateHook implements ORecordHook {
  @Override
  public void onUnregister() {
  }

  @Override
  public RESULT onTrigger(TYPE iType, ORecord iRecord) {
    if (iType.equals(TYPE.FINALIZE_CREATION) || iType.equals(TYPE.FINALIZE_UPDATE)) {
      if (iRecord instanceof ODocument && ((ODocument) iRecord).getDirtyFields().length == 0) {
        ODocument doc = (ODocument) iRecord;
        VertexState.getInstance().setName(doc.field("name"));
        VertexState.getInstance().setDescription(doc.field("description"));
      }
    }
    return null;
  }

  @Override
  public DISTRIBUTED_EXECUTION_MODE getDistributedExecutionMode() {
    return DISTRIBUTED_EXECUTION_MODE.SOURCE_NODE;
  }

}
```

this make your test case pass, it should be ok for your case, the only issue i can see is that on the call of FINALIZE_* you cannot get the list dirty fields, let me know if this completely solve your problem or you need a better solution.

**Comment by lvca** [comment]
Ok guys, if this fixes the problem I'd move this issue to 2.2 together with the new dirty manager that will clean up such things.

**Comment by ylopotun** [comment]
Thanks for the work-around, but we rely on dirty fields, so it won't completely solve our problem... It's not a blocker for us, though, so we can wait for 2.2 for a clean fix.

**Comment by prjhub** [comment]
The issue has been closed because no response has been received after a while. In case the issue is still valid, please reopen it or write a comment that explains why it's still valid and a way to reproduce it. Thank you.

**Comment by lvca** [comment]
Already fixed in v2.2.
